### PR TITLE
Not-where fusion

### PR DESCRIPTION
**Pull request by ashbhandare**
This fusion removes the redundant Not in a Not->Where pattern by swapping the Where inputs 1 and 2.
```
Condition ->  Not -> Where ->
              value0-|  |
              value1----|
gets transformed to:
Condition -> Where ->
      value1-|  |
      value0----|
```

### CODE REVIEW

File: onnxruntime/core/optimizer/not_where_fusion.cc [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+    graph_utils::RemoveNodeOutputEdges(graph, not_node);
+    graph.RemoveNode(not_node.Index());
+
+    graph_utils::ReplaceNodeInput(where_node, 0, *not_input);
+    graph_utils::ReplaceNodeInput(where_node, 1, *where_inputs[2]);
+    graph_utils::ReplaceNodeInput(where_node, 2, *where_inputs[1]);
+  }
+
+  modified = true;
```
Comment (satyajandhyala) [inline onnxruntime/core/optimizer/not_where_fusion.cc:68]: Is this flag set whether or not the graph has Not -> Where pattern?

File: onnxruntime/core/optimizer/not_where_fusion.cc [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+Condition -> Where ->
+      value1-|  |
+      value0----|
+ */
+Status NotWhereFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+  for (auto node_index : node_topology_list) {
```
Comment (satyajandhyala) [inline onnxruntime/core/optimizer/not_where_fusion.cc:29]: Is it possible to iterate using graph.Nodes(), which doesn't require GraphViewer?

File: onnxruntime/test/optimizer/graph_transform_test.cc [hunk 3]
```diff
@@ -527,6 +528,24 @@ TEST_F(GraphTransformationTests, FuseConvBNMulAddUnsqueeze) {
   }
 }
 
+TEST_F(GraphTransformationTests, NotWhereFusion) {
```
Comment (satyajandhyala) [inline onnxruntime/test/optimizer/graph_transform_test.cc:531]: Are there any counter examples to test. For example, Not node output is also a graph output. The code does not fuse Not node and Where node.

File: onnxruntime/core/optimizer/not_where_fusion.cc [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/utils.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace onnxruntime::common;
+namespace onnxruntime {
+
+/**
+Transform that fuses two Not -> Where nodes to a single Where node
```
Comment (satyajandhyala) [inline onnxruntime/core/optimizer/not_where_fusion.cc:15]: What are the preconditions for the transformation?

File: orttraining/orttraining/core/optimizer/graph_transformer_utils.cc [hunk 5]
```diff
@@ -80,6 +81,8 @@ std::vector<std::unique_ptr<GraphTransformer>> GeneratePreTrainingTransformers(
       // Remove duplicate nodes. Must be applied before any recompute transformations.
       transformers.emplace_back(onnxruntime::make_unique<CommonSubexpressionEliminationApplyOnce>(compatible_eps));
 
+      transformers.emplace_back(onnxruntime::make_unique<NotWhereFusion>(compatible_eps));
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/optimizer/graph_transformer_utils.cc:84]: I think this optimization is also harmless to add in the InferenceSession's graph transformer list. 
Please check with "ONNX Runtime Core" alias.

File: onnxruntime/core/optimizer/not_where_fusion.cc [hunk 6]
```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/not_where_fusion.h"
```
Comment (ytaous) [inline onnxruntime/core/optimizer/not_where_fusion.cc:4]: nit: maybe move this line below to be together with other includes from same dir

File: onnxruntime/core/optimizer/not_where_fusion.cc (hunk 1 above)
Comment (ashbhandare) [reply onnxruntime/core/optimizer/not_where_fusion.cc:68]: changing to rewrite rule

File: onnxruntime/core/optimizer/not_where_fusion.cc (hunk 2 above)
Comment (ashbhandare) [reply onnxruntime/core/optimizer/not_where_fusion.cc:29]: changing to rewrite rule

File: onnxruntime/core/optimizer/not_where_fusion.cc (hunk 4 above)
Comment (ashbhandare) [reply onnxruntime/core/optimizer/not_where_fusion.cc:15]: there are none

File: onnxruntime/core/optimizer/not_where_fusion.cc (hunk 6 above)
Comment (ashbhandare) [reply onnxruntime/core/optimizer/not_where_fusion.cc:4]: this follows Google's cpp styleguide https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: onnxruntime/test/optimizer/graph_transform_test.cc (hunk 3 above)
Comment (ashbhandare) [reply onnxruntime/test/optimizer/graph_transform_test.cc:531]: added to the test graph

**Review by SherlockNoMad — APPROVED** [review]

**Comment by SherlockNoMad** [comment]
merge with latest master should resolve the 'Linux GPU CI Pipeline' failure. 
Fix is from https://github.com/microsoft/onnxruntime/pull/7203

File: orttraining/orttraining/core/optimizer/graph_transformer_utils.cc (hunk 5 above)
Comment (ashbhandare) [reply orttraining/orttraining/core/optimizer/graph_transformer_utils.cc]: I have checked with them, adding

File: onnxruntime/core/optimizer/not_where_fusion.cc [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  if (!graph_utils::CanRemoveNode(graph, *p_not_node, logger)) {
+    return false;
+  }
+
+  return true;
+}
+
+Status NotWhereFusion::Apply(Graph& graph, Node& node, RewriteRuleEffect& rule_effect, const logging::Logger&) const {
```
Comment (ytaous) [inline onnxruntime/core/optimizer/not_where_fusion.cc:46]: >Apply [](start = 23, length = 5)

ApplyImpl?

Comment (ashbhandare) [reply onnxruntime/core/optimizer/not_where_fusion.cc:46]: no as this is a RewriteRule

**Review by ytaous — APPROVED** [review]

File: onnxruntime/core/optimizer/not_where_fusion.cc [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+#include "core/optimizer/utils.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace onnxruntime::common;
+namespace onnxruntime {
+
+/**
+Transform that fuses two Not -> Where nodes to a single Where node
+with the where inputs 1 and 2 flipped.
+Condition ->  Not -> Where ->
```
Comment (ytaous) [inline onnxruntime/core/optimizer/not_where_fusion.cc:17]: kindly add multiple where condition here

File: onnxruntime/core/optimizer/not_where_fusion.cc [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+      // Make sure the two nodes do not span execution providers.
+      p_not_node->GetExecutionProviderType() != node.GetExecutionProviderType()) {
+    return false;
+  }
+
+  if (p_not_node->GetOutputEdgesCount() > 1) {
+    // all consumers of not must be where
+    for (auto it = p_not_node->OutputNodesBegin(); it != p_not_node->OutputNodesEnd(); ++it) {
+      if (!graph_utils::IsSupportedOptypeVersionAndDomain(*it, "Where", {9})) {
+        return false;
```
Comment (ytaous) [inline onnxruntime/core/optimizer/not_where_fusion.cc:42]: add one more UT for this ?

Comment (ashbhandare) [reply onnxruntime/core/optimizer/not_where_fusion.cc:42]: this single UT graph has all of these cases

**Review by ytaous — APPROVED** [review]

**Comment by SherlockNoMad** [comment]
Thanks @ashbhandare for the fix! 
Thanks @ytaous for the prompt review!

**Pull request merged**